observability/src/main/java/io/grpc/observability/logging/GcpLogSink.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2022 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.observability.logging;
+
+import com.google.cloud.MonitoredResource;
+import com.google.cloud.logging.LogEntry;
+import com.google.cloud.logging.Logging;
+import com.google.cloud.logging.LoggingOptions;
+import com.google.cloud.logging.Payload.JsonPayload;
+import com.google.cloud.logging.Severity;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.protobuf.util.JsonFormat;
+import io.grpc.internal.JsonParser;
+import io.grpc.observabilitylog.v1.GrpcLogRecord;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Sink for Google Cloud Logging.
+ */
+public class GcpLogSink implements Sink {
+  private final Logger logger = Logger.getLogger(GcpLogSink.class.getName());
+
+  // TODO (dnvindhya): Make cloud logging service a configurable value
+  private static final String SERVICE_TO_EXCLUDE = "google.logging.v2.LoggingServiceV2";
+  private static final String DEFAULT_LOG_NAME = "grpc";
+  private final Logging gcpLoggingClient;
+
+  private static Logging createLoggingClient(String projectId) {
+    LoggingOptions.Builder builder = LoggingOptions.newBuilder();
+    if (!Strings.isNullOrEmpty(projectId)) {
+      builder.setProjectId(projectId);
+    }
+    return builder.build().getService();
+  }
+
+  /**
+   * Retrieves a single instance of GcpLogSink.
+   *
+   * @param destinationProjectId cloud project id to write logs
+   */
+  public GcpLogSink(String destinationProjectId) {
+    this(createLoggingClient(destinationProjectId));
+  }
+
+  @VisibleForTesting
+  GcpLogSink(Logging client) {
+    this.gcpLoggingClient = client;
+  }
+
+  @Override
+  public void write(GrpcLogRecord logProto) {
+    if (gcpLoggingClient == null) {
+      logger.log(Level.SEVERE, "Attempt to write after GcpLogSink is closed.");
+      return;
+    }
+    if (SERVICE_TO_EXCLUDE.equals(logProto.getServiceName())) {
+      return;
+    }
+    try {
+      Map<String, Object> mapPayload = protoToMapConverter(logProto);
+      Severity logEntrySeverity = getCloudLoggingLevel(logProto.getLogLevel());
+      // TODO(vindhyan): make sure all (int, long) values are not displayed as double
+      LogEntry grpcLogEntry =
+          LogEntry.newBuilder(JsonPayload.of(mapPayload))
```

Comment:
nit: just inline mapPayload i.e. replace this with the expression on RHS on line 79

Reply:
Done